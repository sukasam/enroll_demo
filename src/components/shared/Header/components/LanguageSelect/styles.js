import { css } from "@emotion/react";

export default {
    container: css`
        .MuiFormControl-root {
            margin: 0;
        }
    `,

    select: theme => css`
        .MuiOutlinedInput-notchedOutline {
            display: none;
        }

        .MuiInputBase-root {
            font-family: ${theme.typography.fontFamily};
            font-weight: 400;
            font-size: ${theme.typography.fontSize};
            color: ${theme.palette.text.secondary};
            margin-top: ${theme.spacing(0.25)}; // 2px
            min-width: 50px;
            width: 50px;

            @media (min-width: ${theme.breakpoints.values.sm}px) {
                min-width: 120px;
                width: 120px;
            }
        }

        .MuiSelect-select {
            padding-right: 24px !important;
        }
    `
};
