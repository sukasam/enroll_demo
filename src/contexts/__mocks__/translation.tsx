import { Alpha2 } from "Constants/countryConfig/enums.js";

export const useTranslations = jest.fn().mockReturnValue({
    country: "US" as Alpha2,
    language: "en",
    setCountry: jest.fn(),
    setLanguage: jest.fn(),
    fetchTranslations: jest.fn(),
    translations: {}
});

export default useTranslations;
