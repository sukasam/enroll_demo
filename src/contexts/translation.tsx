import { debounce } from "@mui/material";
import { translationsSvc } from "Hydra/translationsSvc";
import { sendEvent } from "Services/googleAnalytics";
import { getCookie, setCookie } from "cookies-next";
import {
    ReactNode,
    createContext,
    useCallback,
    useContext,
    useEffect,
    useMemo,
    useState
} from "react";

export interface TranslationContextType {
    translations: Record<string, string>;
    basicTranslations: Record<string, string>;
    country: string;
    language: string;
    setTranslations: (translations: Record<string, string>) => void;
    setBasicTranslations: (basicTranslations: Record<string, string>) => void;
    setCountry: (country: string) => void;
    setLanguage: (language: string) => void;
    fetchTranslations: (country: string, language: string) => void;
}

const defaultState: TranslationContextType = {
    translations: {},
    basicTranslations: {},
    country: getCookie("country") || "US",
    language: getCookie("language") || "en",
    setTranslations: () => null,
    setBasicTranslations: () => null,
    setCountry: () => null,
    setLanguage: () => null,
    fetchTranslations: () => null
};

function getInitialState(): TranslationContextType {
    if (typeof window !== "undefined") {
        const cachedData = sessionStorage.getItem("translationContext");
        if (cachedData) {
            try {
                return JSON.parse(cachedData);
            } catch (error) {
                console.error(
                    "Failed to parse translation context data from sessionStorage:",
                    error
                );
            }
        }
    }
    return defaultState;
}

const TranslationContext = createContext<TranslationContextType>(defaultState);
TranslationContext.displayName = "Translations Context";

function HtmlLangSync(props: { language: string }): React.ReactElement | null {
    const { language } = props;
    useEffect(() => {
        if (typeof document !== "undefined" && language) {
            document.documentElement.lang = language;
        }
    }, [language]);
    return null;
}

export function TranslationProvider({
    children,
    testValues
}: {
    children: ReactNode;
    testValues?: Record<string, string>;
}): React.ReactElement {
    const initialState = useMemo(() => getInitialState(), []);

    const [translations, setTranslations] = useState<Record<string, string>>(
        initialState.translations
    );
    const [basicTranslations, setBasicTranslations] = useState<
        Record<string, string>
    >(initialState.basicTranslations);
    const [country, setCountry] = useState<string>(initialState.country);
    const [language, setLanguage] = useState<string>(initialState.language);

    useEffect(() => {
        setCookie("country", country);
        setCookie("language", language);
        sendEvent("page_configration", {
            country,
            language
        });
    }, [country, language]);

    const fetchTranslations = debounce(
        async (_country: string, _language: string) => {
            try {
                const response = await translationsSvc.get({
                    country: _country || country,
                    language: _language || language
                });

                if (response.ok === false) {
                    throw new Error("Failed to fetch translations");
                }
                // const translations = await response.json();
                setTranslations(response);
            } catch (error) {
                console.error("Error fetching translations:", error);
            }
        },
        500
    );

    const updateCachedTranslationData = useCallback(() => {
        if (typeof window !== "undefined") {
            const dataToCache = {
                translations,
                basicTranslations,
                country,
                language
            };
            try {
                sessionStorage.setItem(
                    "translationContext",
                    JSON.stringify(dataToCache)
                );
            } catch (error) {
                console.error(
                    "Failed to save translation context data to sessionStorage:",
                    error
                );
            }
        }
    }, [translations, basicTranslations, country, language]);

    useEffect(() => {
        updateCachedTranslationData();
    }, [updateCachedTranslationData]);

    const value = useMemo(
        () => ({
            translations,
            basicTranslations,
            country,
            language,
            setTranslations,
            setBasicTranslations,
            setCountry,
            setLanguage,
            fetchTranslations
        }),
        [translations, basicTranslations, country, language, fetchTranslations]
    );

    const providerValue = useMemo(
        () => ({
            ...value,
            ...testValues
        }),
        [value, testValues]
    );

    return (
        <TranslationContext.Provider value={providerValue}>
            <HtmlLangSync language={language} />
            {children}
        </TranslationContext.Provider>
    );
}

export const useTranslations = (): TranslationContextType =>
    useContext(TranslationContext);
