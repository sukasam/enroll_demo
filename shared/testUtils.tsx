import { ThemeProvider } from "@mui/material/styles";
import { render, RenderOptions } from "@testing-library/react";
import { lightTheme } from "Styles/theme/themes";

const customRender = (
    ui: React.ReactElement,
    options?: Omit<RenderOptions, "wrapper">
): ReturnType<typeof render> =>
    render(ui, {
        wrapper: ({ children }) => (
            <ThemeProvider theme={lightTheme}>{children}</ThemeProvider>
        ),
        ...options
    });

export * from "@testing-library/react";
export { customRender as render };
