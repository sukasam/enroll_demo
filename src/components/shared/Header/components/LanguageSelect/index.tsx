/** @jsxImportSource @emotion/react */
import { useMediaQuery } from "@mui/material";
import FormControl from "@mui/material/FormControl";
import MenuItem from "@mui/material/MenuItem";
import Select from "@mui/material/Select";
import { useTheme } from "@mui/material/styles";
import useAvailableLanguages from "Components/pages/Home/components/HomeContent/Form/hooks";
import { Alpha2 } from "Constants/countryConfig/enums";
import { useTranslations } from "Contexts/translation";
import mixpanelService from "Services/mixpanel/initializeMixPanel";
import MixpanelEvent from "Services/mixpanel/mixpanelEvent";
import { useRouter } from "next/router";
import { SetStateAction, useEffect } from "react";
import styles from "./styles";

type Language = {
    label: string;
    value: string;
};

const AUTH_ROUTES = ["/login", "/register", "/thank-you"];

export default function LanguageSelect(): JSX.Element | null {
    const { pathname } = useRouter();
    const { language, setLanguage, country, fetchTranslations } =
        useTranslations();
    const theme = useTheme();
    const isMobile = useMediaQuery(theme.breakpoints.down("sm"));
    const isAuthRoute = AUTH_ROUTES.includes(pathname);
    const isLoginRoute = pathname === "/login";

    const availableLanguages = useAvailableLanguages(
        isLoginRoute ? null : (country as Alpha2)
    );

    // Ensure the current language is valid when available languages change
    useEffect(() => {
        if (availableLanguages.length > 0) {
            const isValidLanguage = availableLanguages.some(
                lang => lang.value === language
            );
            if (!isValidLanguage) {
                // Set to the first available language if current is invalid
                setLanguage(availableLanguages[0].value);
            }
        }
    }, [availableLanguages, language, setLanguage]);

    if (!isAuthRoute || availableLanguages.length === 0) {
        return null;
    }

    const updateLanguage = (e: {
        target: { value: SetStateAction<string> };
    }): void => {
        const newLanguage = e.target.value as string;
        const selectedLanguage = availableLanguages.find(
            lang => lang.value === newLanguage
        );

        setLanguage(newLanguage);
        fetchTranslations(
            isLoginRoute ? selectedLanguage?.defaultCountry || "" : country,
            newLanguage
        );

        // Track language change in analytics
        mixpanelService.registerSuperProperties({
            user_language: newLanguage
        });
        mixpanelService.trackEvent(MixpanelEvent.USER_LANGUAGE_CHANGED, {
            previous_user_language: language,
            user_language: newLanguage,
            event_location: window.location.pathname.split("/")[1]
        });
    };

    return (
        <div
            className="language-select"
            data-testid="language_select_container"
        >
            <FormControl
                variant="outlined"
                sx={{
                    m: 1,
                    minWidth: { xs: 50, sm: 120 },
                    top: "0px"
                }}
                size="small"
            >
                <Select
                    css={styles.select}
                    data-testid="language_select"
                    labelId="languageSelect"
                    id="languageSelect"
                    value={language}
                    label="language"
                    onChange={updateLanguage}
                >
                    {availableLanguages.map((language: Language) => (
                        <MenuItem key={language.value} value={language.value}>
                            {isMobile
                                ? language.label.slice(0, 2)
                                : language.label}
                        </MenuItem>
                    ))}
                </Select>
            </FormControl>
        </div>
    );
}
