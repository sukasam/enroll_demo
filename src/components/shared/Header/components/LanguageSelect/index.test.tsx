import "@testing-library/jest-dom";
import useAvailableLanguages from "Components/pages/Home/components/HomeContent/Form/hooks";
import { useTranslations } from "Contexts/translation";
import { useRouter } from "next/router";
import mixpanelService from "Services/mixpanel/initializeMixPanel";
import { fireEvent, render, screen, within } from "Shared/testUtils";
import LanguageSelect from "./index";

// Mock the dependencies
jest.mock("next/router", () => ({
    useRouter: jest.fn()
}));

jest.mock("Contexts/translation", () => ({
    useTranslations: jest.fn()
}));

jest.mock("Components/pages/Home/components/HomeContent/Form/hooks", () => ({
    __esModule: true,
    default: jest.fn()
}));

jest.mock("Services/mixpanel/initializeMixPanel", () => ({
    registerSuperProperties: jest.fn(),
    trackEvent: jest.fn()
}));

describe("LanguageSelect", () => {
    beforeEach(() => {
        (useRouter as jest.Mock).mockReturnValue({ pathname: "/login" });
        (useTranslations as jest.Mock).mockReturnValue({
            language: "en",
            setLanguage: jest.fn(),
            country: "US",
            fetchTranslations: jest.fn()
        });
        (useAvailableLanguages as jest.Mock).mockReturnValue([
            { label: "English", value: "en", defaultCountry: "US" },
            { label: "Spanish", value: "es", defaultCountry: "US" }
        ]);
    });

    it("renders on login, register, and thank-you pages", () => {
        const { rerender } = render(<LanguageSelect />);
        expect(screen.getByRole("combobox")).toBeInTheDocument();

        (useRouter as jest.Mock).mockReturnValue({ pathname: "/register" });
        rerender(<LanguageSelect />);
        expect(screen.getByRole("combobox")).toBeInTheDocument();

        (useRouter as jest.Mock).mockReturnValue({ pathname: "/thank-you" });
        rerender(<LanguageSelect />);
        expect(screen.getByRole("combobox")).toBeInTheDocument();
    });

    it("does not render on other pages", () => {
        (useRouter as jest.Mock).mockReturnValue({ pathname: "/home" });
        const { container } = render(<LanguageSelect />);
        expect(container.firstChild).toBeNull();
    });

    it("displays available languages", () => {
        (useRouter as jest.Mock).mockReturnValue({ pathname: "/register" });
        render(<LanguageSelect />);
        const selectElement = screen.getByRole("combobox");
        fireEvent.mouseDown(selectElement);

        // Use `within` to search for options within the opened dropdown
        const listbox = screen.getByRole("listbox");
        expect(within(listbox).getByText("English")).toBeInTheDocument();
        expect(within(listbox).getByText("Spanish")).toBeInTheDocument();
    });

    it("displays available languages on Login", () => {
        (useRouter as jest.Mock).mockReturnValue({ pathname: "/login" });
        (useAvailableLanguages as jest.Mock).mockReturnValue([
            { label: "English", value: "en", defaultCountry: "US" },
            { label: "Spanish", value: "es", defaultCountry: "US" },
            { label: "French", value: "fr", defaultCountry: "CA" }
        ]);

        render(<LanguageSelect />);
        const selectElement = screen.getByRole("combobox");
        fireEvent.mouseDown(selectElement);

        // Use `within` to search for options within the opened dropdown
        const listbox = screen.getByRole("listbox");
        expect(within(listbox).getByText("English")).toBeInTheDocument();
        expect(within(listbox).getByText("Spanish")).toBeInTheDocument();
        expect(within(listbox).getByText("French")).toBeInTheDocument();
    });

    it("calls setLanguage and fetchTranslations when language is changed", () => {
        const { setLanguage, fetchTranslations } = useTranslations() as any;
        render(<LanguageSelect />);
        const selectElement = screen.getByRole("combobox");
        fireEvent.mouseDown(selectElement);

        const listbox = screen.getByRole("listbox");
        fireEvent.click(within(listbox).getByText("Spanish"));

        expect(setLanguage).toHaveBeenCalledWith("es");
        expect(fetchTranslations).toHaveBeenCalledWith("US", "es");
    });

    it("tracks language change event in mixpanel", () => {
        render(<LanguageSelect />);
        const selectElement = screen.getByRole("combobox");
        fireEvent.mouseDown(selectElement);

        const listbox = screen.getByRole("listbox");
        fireEvent.click(within(listbox).getByText("Spanish"));

        expect(mixpanelService.registerSuperProperties).toHaveBeenCalledWith({
            user_language: "es"
        });
        expect(mixpanelService.trackEvent).toHaveBeenCalledWith(
            "enr_user_language_changed",
            expect.objectContaining({
                previous_user_language: "en",
                user_language: "es"
            })
        );
    });

    it("matches snapshot", () => {
        const { container } = render(<LanguageSelect />);
        expect(container.firstChild).toMatchSnapshot();
    });
});
