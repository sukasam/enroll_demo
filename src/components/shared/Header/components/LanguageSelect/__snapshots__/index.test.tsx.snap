// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`LanguageSelect matches snapshot 1`] = `
<div
  class="language-select"
  data-testid="language_select_container"
>
  <div
    class=MuiFormControl-root css-XXXXX-MuiFormControl-root
  >
    <div
      class=MuiInputBase-root MuiOutlinedInput-root MuiInputBase-colorPrimary MuiInputBase-formControl MuiInputBase-sizeSmall css-XXXXX-MuiInputBase-root-MuiOutlinedInput-root-MuiSelect-root-select
      data-testid="language_select"
    >
      <div
        aria-controls=":r5:"
        aria-expanded="false"
        aria-haspopup="listbox"
        aria-labelledby="languageSelect languageSelect"
        class=MuiSelect-select MuiSelect-outlined MuiInputBase-input MuiOutlinedInput-input MuiInputBase-inputSizeSmall css-XXXXX-MuiSelect-select-MuiInputBase-input-MuiOutlinedInput-input
        id="languageSelect"
        role="combobox"
        tabindex="0"
      >
        English
      </div>
      <input
        aria-hidden="true"
        aria-invalid="false"
        class=MuiSelect-nativeInput css-XXXXX-MuiSelect-nativeInput
        tabindex="-1"
        value="en"
      />
      <svg
        aria-hidden="true"
        class=MuiSvgIcon-root MuiSvgIcon-fontSizeMedium MuiSelect-icon MuiSelect-iconOutlined css-XXXXX-MuiSvgIcon-root-MuiSelect-icon
        data-testid="ArrowDropDownIcon"
        focusable="false"
        viewBox="0 0 24 24"
      >
        <path
          d="M7 10l5 5 5-5z"
        />
      </svg>
      <fieldset
        aria-hidden="true"
        class=MuiOutlinedInput-notchedOutline css-XXXXX-MuiOutlinedInput-notchedOutline
      >
        <legend
          class="css-14lo706"
        >
          <span>
            language
          </span>
        </legend>
      </fieldset>
    </div>
  </div>
</div>
`;
